proposals/0000-scoped-type-variable-puns.md

```diff
@@ -0,0 +1,137 @@
+---
+author: Georgi Lyubenov
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/507).
+
+# Allow puns to bind type variables
+
+The current language extensions `NamedFieldPuns` and `ScopedTypeVariables` have a (surprisingly) missing interaction:
+
+```haskell
+data Thing a = MkThing {field :: a}
+
+f :: Thing a -> a
+f MkThing {field :: x} = undefined :: x
+```
+
+This currently fails to parse. I believe that a natural extension of the interaction the two extensions
+would be to instead have the above mean the same as
+
+```haskell
+data Thing a = MkThing {field :: a}
+
+f :: Thing a -> a
+f MkThing {field = field :: x} = undefined :: x
+```
+
+## Motivation
+
+`NamedFieldPuns` and `ScopedTypeVariables` are both firmly cemented as useful and ubiquitous language extensions.
+As we move towards improved type level programming, we should aim to make the UX as seamless as possible.
+I think that allowing the `MkField {field :: x}` syntax would be considered "obvious" and "natural" to anyone who has used both extensions,
+to the point where when I tried it I had no doubt it would work.
+
+Apart from the obvious convenience that this would provide, it is also less error-prone(as is usual with using field names instead of positional matches) compared to binding type variables in patterns, as described [here](https://gitlab.haskell.org/ghc/ghc/-/issues/18830#note_308891).
+
+## Proposed Change Specification
+
+I propose that a pattern match of the form
+```haskell
+Foo {bar :: baz}
+```
+be parsed successfully and desugared(in the renamer) to
+```haskell
+Foo {bar = bar :: baz}
+```
+
+Example grammar change from `ghc`'s `Parser.y`:
+Current:
+```
+fbind
+        : qvar '=' texp
+        | qvar
+        ...
+```
+Proposed addition:
+```
+fbind
+        : qvar '=' texp
+        | qvar '::' ctype
+        | qvar
+        ...
+```
+
+TODO: I'm not sure if the Haskell report contains a specification for the grammar of field puns and scoped type variables.
+If anyone wants to point me to where those are specified, I'd be happy to update this section.
+
+TODO: I'm not sure what corner cases might exist for this, since it is only a syntax sugar change.
```

Comment:
I'm good at corner cases, but I'm struggling here too. Haskell does allow `data Rec = MkRec { a,b,c :: Int }`, and presumably you would not allow `f (MkRec { a,b,c :: x }) = ...`, (or, rather, you would, but the type signature would connect only to the final field, `c`), but that's not too cornerish.

Reply:
Thanks, I noted that down.